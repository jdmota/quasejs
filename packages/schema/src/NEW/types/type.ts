import { type Opaque } from "../../../../util/miscellaneous";
import { SchemaOpCtx } from "../util/context";
import {
  then,
  type ValidationOK,
  type ValidationResult,
  type ValidationResultMaybeAsync,
} from "../util/result";
import type { JsType } from "./js-types";

export type RefinerOpts = Partial<{
  readonly message: string;
  readonly path: readonly string[];
}>;

export type SchemaStatic<S> = S extends SchemaType<infer T> ? T : never;

export abstract class SchemaType<T> {
  test(value: unknown): value is T {
    const result = this.validate(value, new SchemaOpCtx());
    if (result instanceof Promise) {
      throw new Error("test() should be sync");
    }
    return result.ok;
  }

  abstract validate(
    value: unknown,
    ctx?: SchemaOpCtx
  ): ValidationResultMaybeAsync<T>;

  refine(
    fn: (value: T) => boolean | Promise<boolean>,
    opts: RefinerOpts = {}
  ): SchemaType<T> {
    return new SchemaRefinementSimpleType(this, fn, opts);
  }

  refineMore(
    fn: (value: T, ctx: SchemaOpCtx) => void | Promise<void>
  ): SchemaType<T> {
    return new SchemaRefinementMoreType(this, fn);
  }

  opaque<B extends PropertyKey>(brand: B): SchemaType<Opaque<T, B>> {
    return new SchemaOpaqueType<T, B>(this, brand);
  }

  cast<C>(type: JsType<C>): SchemaType<C> {
    return new SchemaCastType<C, T>(this, type);
  }
}

export class SchemaRefinementSimpleType<T> extends SchemaType<T> {
  constructor(
    private readonly parent: SchemaType<T>,
    private readonly fn: (value: T) => boolean | Promise<boolean>,
    private readonly opts: RefinerOpts = {}
  ) {
    super();
  }

  override validate(
    value: unknown,
    ctx = new SchemaOpCtx()
  ): ValidationResultMaybeAsync<T> {
    return then(this.parent.validate(value, ctx), r => {
      if (r.ok) {
        return then(this.fn(r.value), bool => {
          if (!bool) {
            if (this.opts.path) for (const p of this.opts.path) ctx.push(p);
            ctx.addError(this.opts.message ?? "Validation error");
          }
          return ctx.result(r.value);
        });
      }
      return r;
    });
  }
}

export class SchemaRefinementMoreType<T> extends SchemaType<T> {
  constructor(
    private readonly parent: SchemaType<T>,
    private readonly fn: (value: T, ctx: SchemaOpCtx) => void | Promise<void>
  ) {
    super();
  }

  override validate(
    value: unknown,
    ctx = new SchemaOpCtx()
  ): ValidationResultMaybeAsync<T> {
    return then(this.parent.validate(value, ctx), r => {
      if (r.ok) {
        return then(this.fn(r.value, ctx), () => ctx.result(r.value));
      }
      return r;
    });
  }
}

export class SchemaOpaqueType<T, B extends PropertyKey> extends SchemaType<
  Opaque<T, B>
> {
  constructor(
    private readonly parent: SchemaType<T>,
    public readonly brand: B
  ) {
    super();
  }

  override validate(
    value: unknown,
    ctx = new SchemaOpCtx()
  ): ValidationResultMaybeAsync<Opaque<T, B>> {
    return then(this.parent.validate(value, ctx), r =>
      r.ok ? (r satisfies ValidationResult<T> as ValidationOK<Opaque<T, B>>) : r
    );
  }
}

export class SchemaCastType<T, I = T> extends SchemaType<T> {
  constructor(
    private readonly parent: SchemaType<I>,
    public readonly jsType: JsType<T>
  ) {
    super();
  }

  override validate(
    value: unknown,
    ctx = new SchemaOpCtx()
  ): ValidationResultMaybeAsync<T> {
    return this.parent.validate(
      value,
      ctx
    ) satisfies ValidationResultMaybeAsync<I> as ValidationResultMaybeAsync<T>;
  }
}
