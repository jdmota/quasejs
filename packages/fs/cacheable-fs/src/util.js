// @flow

const pathUrl = require( "@quase/path-url" );
const path = require( "path" );

export function dirname( p: string ): string {
  return pathUrl.lowerPath( path.dirname( p ) );
}

export function makeAbsolutePath( p: string ): string {
  return pathUrl.lowerPath( pathUrl.makeAbsolutePath( p ) );
}
