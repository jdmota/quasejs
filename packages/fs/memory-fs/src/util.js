const pathUrl = require( "@quase/path-url" );
const path = require( "path" );

function lower( p ) {
  return path.sep === "\\" ? p.toLowerCase() : p;
}

export function dirname( p ) {
  return lower( path.dirname( p ) );
}

export function makeAbsolute( p ) {
  return pathUrl.isUrl( p ) ? pathUrl.makeAbsoluteUrl( p ) : makeAbsolutePath( p );
}

export function makeAbsolutePath( p ) {
  return lower( pathUrl.makeAbsolutePath( p ) );
}

export function makeAbsoluteUrl( p ) {
  return pathUrl.makeAbsoluteUrl( p );
}

export const isUrl = pathUrl.isUrl;
