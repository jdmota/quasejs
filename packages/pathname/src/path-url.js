import req from "../../_helper/require";

const path = req( "path" );
const URL = typeof window !== "undefined" && window.URL; // eslint-disable-line no-undef
const LOCATION = typeof window !== "undefined" && window.location; // eslint-disable-line no-undef

export function makeAbsolute( file ) {
  return URL ? new URL( file, LOCATION ).href : path.resolve( file );
}

export function resolveAsUrl( from, to ) {
  return URL ? new URL( to, from ).href : path.resolve( path.dirname( from ), to );
}

export function normalize( file ) {
  if ( URL ) {
    return new URL( file, LOCATION ).pathname;
  }
  return path.relative( process.cwd(), file ).replace( /\\/g, "/" );
}
