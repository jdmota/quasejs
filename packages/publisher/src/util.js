require( "any-observable/register/rxjs-all" ); // eslint-disable-line import/no-unassigned-import
const { merge } = require( "rxjs" );
const { filter } = require( "rxjs/operators" );
const streamToObservable = require( "@samverschueren/stream-to-observable" );
const split = require( "split" );
const execa = require( "execa" );
const Listr = require( "listr" );
const issueRegex = require( "issue-regex" );
const terminalLink = require( "terminal-link" );

export function exec( cmd, args, opts ) {
  const cp = execa( cmd, args, opts );

  return merge(
    streamToObservable( cp.stdout.pipe( split() ), { await: cp } ),
    streamToObservable( cp.stderr.pipe( split() ), { await: cp } )
  ).pipe( filter( Boolean ) );
}

export function l( tasks ) {
  return new Listr( tasks || [], {
    // showSubtasks: false
  } );
}

export function error( message ) {
  const err = new Error( message );
  err.__generated = true;
  return err;
}

export function linkifyCompare( url, a, b ) {
  if ( !url ) {
    return `${a}...${b}`;
  }
  if ( !terminalLink.isSupported ) {
    return `${url}/compare/${a}...${b}`;
  }
  return terminalLink( `${a}...${b}`, `${url}/compare/${a}...${b}` );
}

// Adapted from https://github.com/sindresorhus/np

export function linkifyIssues( url, message ) {
  if ( !url || !terminalLink.isSupported ) {
    return message;
  }
  return message.replace( issueRegex(), issue => {
    const issuePart = issue.replace( "#", "/issues/" );
    if ( issue.startsWith( "#" ) ) {
      return terminalLink( issue, `${url}${issuePart}` );
    }
    const host = url.replace( /\.com\/[^]*$/, ".com/" );
    return terminalLink( issue, `${host}${issuePart}` );
  } );
}

export function linkifyCommit( url, commit ) {
  if ( !url || !terminalLink.isSupported ) {
    return commit;
  }
  return terminalLink( commit, `${url}/commit/${commit}` );
}
