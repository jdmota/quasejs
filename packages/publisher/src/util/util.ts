import execa from "execa";
import logSymbols from "log-symbols";
import terminalLink from "terminal-link";
import issueRegex from "issue-regex";
import History from "../history";

const { filter } = require( "rxjs/operators" );
const streamToObservable = require( "@samverschueren/stream-to-observable" );
const split = require( "split" );
const Listr = require( "listr" );

type ExecOpts = {
  cwd?: string;
  history?: History;
  preferLocal?: boolean;
};

export function errorIgnoreStack( error: any ) {
  error.__generated = true;
  throw error;
}

export async function execPromise( cmd: string, args: string[], opts: ExecOpts = {} ) {
  const { history } = opts;
  const operation = [ cmd ].concat( args );
  try {
    if ( history ) history.start( operation );
    const cp = await execa( cmd, args, {
      preferLocal: true,
      ...opts
    } );
    if ( history ) history.end( operation );
    return cp;
  } catch ( error ) {
    error.__generated = true;
    throw error;
  }
}

export async function execStdout( cmd: string, args: string[], opts: ExecOpts = {} ): Promise<string> {
  const { history } = opts;
  const operation = [ cmd ].concat( args );
  try {
    if ( history ) history.start( operation );
    const { stdout } = await execa( cmd, args, {
      preferLocal: true,
      ...opts
    } );
    if ( history ) history.end( operation );
    return stdout;
  } catch ( error ) {
    error.__generated = true;
    throw error;
  }
}

export function execObservable( cmd: string, args: string[], opts: ExecOpts = {} ) {
  const { history } = opts;
  const operation = [ cmd ].concat( args );
  const cp = execa( cmd, args, {
    preferLocal: true,
    ...opts
  } );

  if ( history ) {
    history.start( operation );
    cp.then( () => history.end( operation ) );
  }

  const { all } = cp as any;
  return streamToObservable( all.pipe( split() ) ).pipe( filter( Boolean ) );
}

export function l( tasks?: any[] ) {
  return new Listr( tasks || [], {
    // showSubtasks: false
  } );
}

export function error( message: string ) {
  const err = new Error( message );
  // @ts-ignore
  err.__generated = true;
  return err;
}

export function linkifyCompare( url: string, a: string, b: string ): string {
  if ( !url ) {
    return `${a}...${b}`;
  }
  if ( !terminalLink.isSupported ) {
    return `${url}/compare/${a}...${b}`;
  }
  return terminalLink( `${a}...${b}`, `${url}/compare/${a}...${b}` );
}

// Adapted from https://github.com/sindresorhus/np

export function linkifyIssues( url: string, message: string ): string {
  if ( !url || !terminalLink.isSupported ) {
    return message;
  }
  return message.replace( issueRegex(), issue => {
    const issuePart = issue.replace( "#", "/issues/" );
    if ( issue.startsWith( "#" ) ) {
      return terminalLink( issue, `${url}${issuePart}` );
    }
    const host = url.replace( /\.com\/[^]*$/, ".com/" );
    return terminalLink( issue, `${host}${issuePart}` );
  } );
}

export function linkifyCommit( url: string, commit: string ): string {
  if ( !url || !terminalLink.isSupported ) {
    return commit;
  }
  return terminalLink( commit, `${url}/commit/${commit}` );
}

export function linkifyCommitRange( url: string, commitRange: string ): string {
  if ( !url || !terminalLink.isSupported ) {
    return commitRange;
  }
  return terminalLink( commitRange, `${url}/commit/${commitRange}` );
}

export function info( message: string ) {
  // eslint-disable-next-line no-console
  console.log( `${logSymbols.info} ${message}` );
}
