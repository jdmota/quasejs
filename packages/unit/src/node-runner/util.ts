import turbocolor from "turbocolor";

export function setExitCode(code: number) {
  if (!process.exitCode) {
    process.exitCode = code;
  }
}

export function validationError(message: string) {
  const err: any = new Error(message);
  err.__validation = true;
  return err;
}

export function printError(error: Error) {
  let message;
  // @ts-ignore
  if (error.__validation || !error.stack) {
    message = `${turbocolor.bold(
      "Validation Error"
    )}:\n\n${error.message.replace(/^(?!$)/gm, "  ")}`;
  } else {
    message = error.stack;
  }
  console.error(`${turbocolor.red(message)}\n`); // eslint-disable-line no-console
}
