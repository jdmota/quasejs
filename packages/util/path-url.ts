import path from "node:path";
import url, { fileURLToPath } from "node:url";
import slash from "slash";
import urlRegexDefault from "url-regex";

type Url = string;
type Path = string;

// File paths

export const pathToUrl = url.pathToFileURL;
export const urlToPath = url.fileURLToPath;

export { slash };

export function isAbsolutePath(file: Path) {
  return path.isAbsolute(file);
}

export function makeAbsolutePath(file: Path) {
  return resolvePath(process.cwd(), file);
}

export function resolvePath(from: Path, to: Path) {
  return path.resolve(from, to);
}

export function resolvePathAsUrl(from: Path, to: Path) {
  return path.resolve(path.dirname(from), to);
}

export function relativePath(id: Path, cwd: Path): Path {
  return path.relative(cwd, id);
}

export function relativePathAsUrl(id: Path, cwd: Path): Path {
  return path.relative(path.dirname(cwd), id);
}

export function prettifyPath(file: Path): string {
  return slash(lowerPath(path.relative(process.cwd(), file)));
}

export function lowerPath(file: Path): Path {
  return path.sep === "\\" ? file.toLowerCase() : file;
}

export function normalizePath(file: Path): Path {
  return slash(lowerPath(makeAbsolutePath(file)));
}

export const reExt = /\.(.+)$/;

export function getType(id: string): string {
  const match = id.match(reExt);
  return match ? match[1] : "";
}

// Receives import.meta.url
export function getDirnameFromMetaUrl(url: string) {
  return path.dirname(fileURLToPath(url));
}

// Urls

const LOCATION =
  (typeof window !== "undefined" &&
    window.location &&
    new URL(window.location.href)) ||
  new URL("http://localhost/");

const reAbsUrl = /^[a-z][a-z0-9+.-]*:/;

export function isAbsoluteUrl(url: Url) {
  return reAbsUrl.test(url);
}

export function makeAbsoluteUrl(url: Url): string {
  return new URL(url, LOCATION).href;
}

export function resolveUrl(from: Url, to: Url): string {
  return new URL(to, from).href;
}

export function prettifyUrl(
  url: Url,
  opts?: { readonly lastSlash: boolean }
): string {
  let { hash, origin, pathname, search, protocol } = new URL(url, LOCATION);
  pathname = opts && opts.lastSlash ? pathname : removeLastSlash(pathname);
  origin = origin === LOCATION.origin ? "" : origin;
  if (protocol === "file:") {
    if (search || hash) {
      return "file://" + pathname + search + hash;
    }
    return prettifyPath(pathname.slice(1)); // Remove the third slash
  }
  return origin + pathname + search + hash;
}

export function removeLastSlash(url: Url) {
  return url.replace(/\/+$/g, "");
}

// Both

const urlRegex = urlRegexDefault({ exact: true });

export function isUrl(url: string) {
  return urlRegex.test(url.trim()) || url.startsWith("file://");
}

export function isAbsolute(name: Url | Path) {
  return isUrl(name) ? isAbsoluteUrl(name) : isAbsolutePath(name);
}

export function makeAbsolute(name: Url | Path) {
  return isUrl(name) ? makeAbsoluteUrl(name) : makeAbsolutePath(name);
}

export function resolve(from: Url | Path, to: Url | Path) {
  return isUrl(from) ? resolveUrl(from, to) : resolvePath(from, to);
}

export function resolveAsUrl(from: Url | Path, to: Url | Path) {
  return isUrl(from) ? resolveUrl(from, to) : resolvePathAsUrl(from, to);
}

export function prettify(p: Url | Path, opts?: { lastSlash: boolean }) {
  return isUrl(p) ? prettifyUrl(p, opts) : prettifyPath(p);
}
