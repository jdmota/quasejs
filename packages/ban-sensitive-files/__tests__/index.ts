import { prettifyPath } from "../../util/path-url";
import { run } from "..";
import type { FileReport } from "../types";

function cleanReport(report: FileReport): FileReport {
  return {
    ...report,
    filename: prettifyPath(report.filename),
  };
}

it("banned files", async () => {
  const results = [];

  for await (const report of run({
    folder: "packages/ban-sensitive-files",
    all: true,
    verbose: true,
  })) {
    if (report.kind !== "ok") {
      results.push(cleanReport(report));
    }
  }

  expect(results).toMatchSnapshot();
});
