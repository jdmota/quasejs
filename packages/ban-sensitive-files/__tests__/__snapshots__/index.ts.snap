// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`banned files 1`] = `
[
  {
    "check": {
      "pattern": "auth"?\\s*(:|=)",
      "type": "regex",
    },
    "filename": "packages/ban-sensitive-files/__tests__/examples/.npmrc",
    "kind": "sensitive",
    "rule": {
      "caption": "config file",
      "checks": [
        {
          "pattern": "auth"?\\s*(:|=)",
          "type": "regex",
        },
        {
          "pattern": "password"?\\s*(:|=)",
          "type": "regex",
        },
        {
          "pattern": "token"?\\s*(:|=)",
          "type": "regex",
        },
      ],
      "part": "filename",
      "pattern": "rc|config",
      "type": "regex",
    },
  },
  {
    "filename": "packages/ban-sensitive-files/__tests__/examples/example.pem",
    "kind": "banned",
    "rule": {
      "caption": "Potential cryptographic private key",
      "description": null,
      "part": "extension",
      "pattern": "pem",
      "type": "match",
    },
  },
]
`;
